setup.py

```diff
@@ -39,19 +40,9 @@ def get_scm_config():
 
     return {} # use the default config
 
-dev_requirements = [
-    "pytest",
-    "pytest-xdist",
-]
-
-setup(
-    name="hpy.devel",
-    packages = ['hpy.devel'],
-    include_package_data=True,
-    extras_require={
-        "dev": dev_requirements,
-    },
-    ext_modules = [
+EXT_MODULES = []
+if sys.implementation.name == 'cpython':
```

Comment:
This is a bit strange because currently this repo is two things:

1) a toolkit of headers, etc for developing hpy extensions and compiling them either as cpython-only or universal
2) an implementation of hpy.universal for cpython.

Perhaps at some point hpy.universal for cpython should be split out into its own section more properly and then it is only a dependency of the C extension that uses the HPy API and not of hpy itself.

Reply:
Do you think there will be 2 packages on PyPI ? A build time dependency (hpy.devel) and a runtime dependency for CPython (hpy.universal) ? I don't know if it is possible (and good) to produce two distinct PyPI packages with one setup.py / pyproject.toml. Also I don't know how to specify a runtime dependency only for one Python implementation.

Reply:
> Also I don't know how to specify a runtime dependency only for one Python implementation.

In a requirements.txt and pyproject.toml you can do

```
hpy.devel ; platform_python_implementation != "PyPy"
```

Reply:
> I don't know if it is possible (and good) to produce two distinct PyPI package

It is non-standard, but has been done in other projects. This was recently discussed on [discuss.python.org](https://discuss.python.org/t/use-case-supported-2-namespace-packages-in-single-source-repo-with-1-setup-py/5440).

Reply:
> This is a bit strange because currently this repo is two things:
[cut]

> Do you think there will be 2 packages on PyPI ? A build time dependency (hpy.devel) and a runtime dependency for CPython (hpy.universal) ? 

yes, the plan since the beginning was to have one repo and two distinct pypi packages (I already pre-registered them on PyPI :)). We need to have them in the same repo because there are some files which are shared between the two so it's much easier to manage this way.

- `hpy.devel` is a build-time dependency. It is supposed to be put in `setup_requires`

- `hpy.universal` is a runtime dependency. For CPython, you will download it from PyPI, but e.g. PyPy will ship its own version (as it does with cffi and vmprof).

> I don't know if it is possible (and good) to produce two distinct PyPI packages with one setup.py / pyproject.toml.

I don't know either. At worst, we can have two different files: `setup_hpy_devel.py` and `setup_hpy_universal.py`: it's a bit nonstandard but I don't see why it shouldn't work.

> Also I don't know how to specify a runtime dependency only for one Python implementation.

PyPy will also ship `hpy.universal.egg-info`, as it [already does](https://foss.heptapod.net/pypy/pypy/-/tree/branch/py3.6/lib_pypy) for `cffi`, `readline`, `greenlet` etc. This way if you do `install_requires=['hpy.universal']`, `pip` and `setuptools` will download the package from PyPI **only** on CPython.

However, there is ANOTHER problem, which might be impossible to solve with the current tools. The dependency on `hpy.universal` is not a property of the package itself, but of the way it is built. If you are on CPython and you build your package with the CPython ABI, the wheel that you produce has no runtime dependency at all. But if you build it with the universal ABI, then you suddenly depends on `hpy.universal`.

I suppose that an easy way to solve is to always depend on `hpy.universal` and if you don't need it... too bad.

Reply:
The last point sounds like [this long discussion](https://discuss.python.org/t/options-to-build-the-same-package-in-different-ways-with-different-build-dependencies/4458) about building the same package in different ways. @paugier it seems you were an active participant there: any more insights after that discussion?

Reply:
I can maybe add that h5py does that to build a standard package built against sequential hdf5 and a MPI package built against hdf5 MPI (see https://github.com/h5py/h5py/pull/1619).

setuptools supports specifying build dependencies both in pyproject.toml and in setup.py, and the later can be controlled with environment variables. It works fine for now with isolated build (we use that in fluiddyn packages) but I guess people working on improving Python packaging don't like that because (from what I understood) (i) it's better to be able to know the build dependencies without Python execution (without setup.py), (ii) there is no mechanism to change the name of wheels depending on the build and to specify that you want to install a specific build (similar to `conda install "blas-devel[build=*openblas]"`).